src/query/shorthand.ts

```diff
@@ -261,3 +264,164 @@ export function fieldDef(encQ: EncodingQuery,
   fieldAndParams += props.map((p) => ',' + p.key + '=' + p.value).join('');
   return (fn ? fn + '(' + fieldAndParams + ')' : fieldAndParams);
 }
+
+export function parse(shorthand: string): SpecQuery {
+  let splitShorthand = shorthand.split('|');
+
+  let specQ: SpecQuery = null;
+
+  // Mark
+  specQ.mark = splitShorthand[0];
+
+  for (let part of splitShorthand) {
+
+    // Encodings
+    for (let channel of [Channel.COLOR, Channel.COLUMN, Channel.DETAIL, Channel.LABEL, Channel.OPACITY,
+                         Channel.ORDER, Channel.PATH, Channel.ROW, Channel.SHAPE, Channel.SIZE,
+                         Channel.TEXT, Channel.X, Channel.X2, Channel.Y, Channel.Y2]) {
+      const prefix = channel + ':';
+      if (part.substr(0, prefix.length) === prefix) {
+        specQ.encodings.push(shorthandParser[channel](part));
+        break;
+      }
+    }
+
+    // Transform
+    for (let transformPart of ['calculate', 'filter', 'filterInvalid']) {
```

Comment:
If something is already a part of the first one, you shouldn't have to loop through here.